content/post/73-putting-npm-on-ipfs-part-1.md

```diff
@@ -0,0 +1,122 @@
+
+Everything is also served over https unlike some very old npm modules which are served over http by the npm registry.
+
+## 👷‍♀️ How we built it
+
+A group of [ipfs-npm-registry-mirror](https://github.com/ipfs-shipyard/ipfs-npm-registry-mirror) nodes are running, they contain a [js-IPFS](https://github.com/ipfs/js-ipfs) instance used to share and resolve modules and are connected to the wider IPFS network.  They also have a small http server used to respond to requests for module metadata and tarballs from the from the npm cli.  These all sit behind an http load balancer to distribute traffic to the mirrors.
+
+![Network topology](/73-putting-npm-on-ipfs-part-1/network-topology.png)
+
+### 🗃️ The datastore
+
+Each mirror has an [IPFS Repository](https://github.com/ipfs/specs/tree/master/repo) that is backed by AWS S3 via an instance of [datastore-s3](https://github.com/ipfs/js-datastore-s3) - this lets us deploy the service as a set of immutable containers, handle huge amounts of data cost effectively and scale up and down as required at the price of a small amount of latency on transfers.
+
+![Datastore S3](/73-putting-npm-on-ipfs-part-1/datastore-s3.png)
+
+If you'd like to leverage S3 for your IPFS node, check out the example of how to configure IPFS to use S3 in the [datastore-s3 examples folder](https://github.com/ipfs/js-datastore-s3/tree/v0.2.3/examples/full-s3-repo).
+
+### 📝 Module metadata
+
+Each module has a set of metadata that describes the versions that are available, along with the tarballs that make up the release.  It is a JSON document that we store in the [MFS](https://docs.ipfs.io/guides/concepts/mfs/) under the directory `/npm-registry`, so, for example for the module [`ipfs`](https://www.npmjs.com/package/ipfs):
+
+```console
+$ jsipfs files read /npm-registry/ipfs
+{"_id":"ipfs","_rev":"122-28686ac76345db3f398b88ae73346a15","name":"ipfs","description":"JavaScript implementation of the IPFS specification","di...
+```
+
+The metadata for a [module on registry.js.ipfs.io](https://registry.js.ipfs.io/ipfs) is identical to that [on the public registry](https://registry.npmjs.org/ipfs), the difference is that we store CIDs and the original download location:
+
+```javascript
+{
+...
+  "name": "ipfs",
+...
+  "versions": {
+    "0.34.4": {
+      "name": "ipfs",
+...
+      "dist": {
+        "tarball": "https://registry.js.ipfs.io/ipfs/-/ipfs-0.34.4.tgz",
+        "source": "https://registry.npmjs.org/ipfs/-/ipfs-0.34.4.tgz",
+        "cid": "bafybeiafts7s65iodk4wsetucwhury3cpk4fge374wxhdu5vzc4zuli4xi"
+      }
+    },
+...
+```
+
+The CID resolves to the tarball for a module, but while the module has been added to IPFS is not stored in the MFS so we resolve it using the CID:
+
+```console
+$ jsipfs files read /ipfs/bafybeiafts7s65iodk4wsetucwhury3cpk4fge374wxhdu5vzc4zuli4xi
```

Comment:
You've just said it's not stored in MFS but have used a command we've documented as being for primary use with MFS to read it - perhaps using `ipfs cat` is more appropriate here?